File: clients/ruby/model/level.rb
require_relative 'tile'
class Level
    attr_accessor :tiles
    def initialize(tiles)
        @tiles = tiles
    end
    def self.read_from(stream)
        tiles = []
        stream.read_int().times do |_|
            tiles_element = []
            stream.read_int().times do |_|
                tiles_element_element = stream.read_int()
                if tiles_element_element < 0 || tiles_element_element > 5
                    raise "Unexpected discriminant value"
                end
                tiles_element.push(tiles_element_element)
            end
            tiles.push(tiles_element)
        end
        Level.new(tiles)
    end
    def write_to(stream)
        stream.write_int(@tiles.length())
        @tiles.each do |element|
            stream.write_int(element.length())
            element.each do |element|
                stream.write_int(element)
            end
        end
    end
end

File: clients/javascript/model/level.js
class Level {
    constructor(tiles) {
        this.tiles = tiles;
    }
    static async readFrom(stream) {
        let tiles;
        tiles = [];
        for (let i = await stream.readInt(); i > 0; i--) {
            let tilesElement;
            tilesElement = [];
            for (let j = await stream.readInt(); j > 0; j--) {
                let tilesElementElement;
                tilesElementElement = stream.readInt();
                tilesElement.push(tilesElementElement);
            }
            tiles.push(tilesElement);
        }
        return new Level(tiles);
    }
    async writeTo(stream) {
        let tiles = this.tiles;
        await stream.writeInt(tiles.length);
        for (let tilesElement of tiles) {
            await stream.writeInt(tilesElement.length);
            for (let tilesElementElement of tilesElement) {
                await stream.writeInt(tilesElementElement);
            }
        }
    }
}
module.exports = { Level: Level }
